### ISSUE TITLE
Inconsistent Behaviour for `.env` File

### ISSUE THREAD

**Issue by berwyn** [docker/compose#6203]
## Description of the issue
I have my app split into multiple compose files so that different devs can run different parts of the stack based on the work they're doing. As an example, a front-end dev who only works on front-end would want to spin up our API, and potentially to a feature branch related to the feature they're working on. To this end, I have something like:

```yaml
# docker-compose.yml
version: '3'
services:
  app:
    # Some config here
```

```yaml
# docker-compose.backend.yml
version: '3'
services:
  api:
    image: ${DOCKER_BACKEND_IMAGE:-myorg/api:staging}
```

and then I have a .env with something like `DOCKER_BACKEND_IMAGE=myorg/api:foo`

When running `docker-compose -f docker-compose.backend.yml up -d` (or other commands like `config` or `images`), the values get templated correctly from `.env` and everything works as expected, but other commands like `run` will instead not use the environment variable and use the default value instead. Is this intended?

## Context information (for bug reports)

```
docker-compose version 1.22.0, build f46880f
docker-py version: 3.4.1
CPython version: 3.6.4
OpenSSL version: OpenSSL 1.0.2o  27 Mar 2018
```

```
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:21:31 2018
 OS/Arch:           darwin/amd64
 Experimental:      true

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:29:02 2018
  OS/Arch:          linux/amd64
  Experimental:     true
 Kubernetes:
  Version:          v1.10.3
  StackAPI:         v1beta2

```

## Steps to reproduce the issue

Assuming you have two tagged images which have differing file contents,

1. Create a compose file with a templated image and default
2. Run `docker-compose up -d` and watch the stack create
3. Use `docker-compose exec` to attach to the service, note that it's the correct one
4. Use `docker-compose run` to run a one-off container, note that it's _not_ the correct image

### Observed result
`up`, `images`, and `config` have the correct values while `run` does not

### Expected result
`run` behaves the same as the other commands

## Additional information
macOS High Sierra with latest patches, Docker for Mac (Community, Edge) installed via the website with Kubernetes enabled but otherwise stock.

I'm more than happy to provide any more clarification or attempt to put a repro together sans the work code if that's helpful, just let me know!

Thanks for your time and assistance :)

**Comment by shin-** [comment]
Hi @berwyn ,

I'm not able to reproduce this locally. `docker-compose run` reads the `.env` file like every other command. Compose is written in a way that configuration parsing and interpolation is always done the same way ahead of any command, so any results that would indicate the contrary are surprising and most likely due to other factors. Are you running all commands in the same environment (environment variables, working directory, user executing the command)?

**Comment by berwyn** [comment]
Hi @shin-,

Sorry for the delay in reply, this slipped under my radar.

I'm running all the `docker-compose` commands within the same shell, one after the other, by the same user, with the same environment. Given that all the parsing is handled the same way, it might be more useful for me to provide a small repro of the setup I'm using. I'll work on creating a minimal repro based on what we're seeing internally at my employer, please give me a day or two to construct that!

**Comment by shin-** [comment]
@berwyn Did you get a chance to create a repro case? Or did you figure out what the issue was?

**Comment by berwyn** [comment]
Sorry, we've got a big project at work and I haven't had as much time as I'd like to build the repro. I've also been struggling to reproduce what causes the issue without the _exact_ conditions I have in this work app, so I wonder if it's just some unicorn of how I use my system and the files as they've been authored.

It may be some time (1 - 2 weeks or so) before I can sit down and properly create a repro of the issue, assuming I don't discover along the way that I've made a bone-headed mistake in authoring the compose files and bash scripts that orchestrate them. That being the case (and since this hasn't seemed to attract any "me too!"s), it might be safe to close this and re-open it if and when I can provide a real reproduction.

So sorry to take your time with this!
